subsys/fs/nvs/nvs.c

```diff
@@ -175,12 +175,51 @@ static int nvs_flash_ate_wrt(struct nvs_fs *fs, const struct nvs_ate *entry)
 }
 
 /* data write */
-static int nvs_flash_data_wrt(struct nvs_fs *fs, const void *data, size_t len)
+static int nvs_flash_data_wrt(struct nvs_fs *fs, const void *data, size_t len, bool compute_crc)
```

Comment:
A little different approach (hope I didn't overlook anything):
```
static int nvs_flash_data_wrt(struct nvs_fs *fs, const void *data, size_t len, bool compute_crc)
{
	int rc;

	/* Only add the CRC if required */
	if (IS_ENABLED(CONFIG_NVS_DATA_CRC) && (compute_crc) && (len != 0U)) {
		const size_t sa_len = len & ~(fs->flash_parameters->write_block_size - 1);
		const size_t rem_len = len - sa_len;
		const uint32_t data_crc = crc32_ieee(data, len);
		uint8_t buf[rem_len + NVS_DATA_CRC_SIZE];
		uint8_t *data8 = (uint8_t *)data;
		
		rc = nvs_flash_al_wrt(fs, fs->data_wra, data8, sa_len);
		fs->data_wra += sa_len;
		if (rc != 0) {
			goto end;
		}


		memcpy(buf, data8 + sa_len, rem_len);
		memcpy(&buf[rem_len], &data_crc, NVS_DATA_CRC_SIZE);
		rc = nvs_flash_al_wrt(fs, fs->data_wra, buf, sizeof(buf));
		fs->data_wra += nvs_al_size(fs, sizeof(buf));
		if (rc != 0) {
			goto end;
		}
	} else {
		rc = nvs_flash_al_wrt(fs, fs->data_wra, data, len);
		fs->data_wra += nvs_al_size(fs, len);
	}

end:
	return rc;
}
```

Reply:
Is there an issue with my current implementation ?

Reply:
> Is there an issue with my current implementation ?

Is is not that there is an issue with the current implementation. I was adding a remark about how to avoid any `#ifdef's` in the code and how to simplify the calculation of `aligned_len` and some other nits that I tought it would be simpler to propose an "alternative" that you could pick idea's from (BTW I think the last `if (rc != 0)` can also be removed).

Reply:
OK, thanks ! I will update the PR with some of your proposals.